src/Aspire.Hosting/Dashboard/DashboardViewModelService.cs

```diff
@@ -294,6 +325,82 @@ private async Task WriteChange(ResourceViewModel resourceViewModel, ObjectChange
             .ConfigureAwait(false);
     }
 
+    // Check a service endpoint
+    private async Task<bool> CheckServiceEndpointReady(string proxyUrlString, string errorContent)
+    {
+        using HttpClient client = new();
+        try
+        {
+            var response = await client.GetAsync(proxyUrlString).ConfigureAwait(false);
+            // Traefik returns a 404 with content length of 19 when it is not ready
+            if (response.StatusCode == HttpStatusCode.NotFound
+                && response.Content.Headers.ContentLength == 19)
+            {
+                // If the response content doesn't exactly match the expected return from traefik, we assume
+                // the endpoint is ready
+                var content = await response.Content.ReadAsStringAsync(_cancellationToken).ConfigureAwait(false);
+                if (!string.Equals(content, errorContent, StringComparison.Ordinal))
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                // Endpoint returned a valid response
+                return true;
+            }
+        }
+        catch
```

Comment:
do we want to retry on other exceptions? This may cause retrying for something which may never succeed but I don't have good ideas what should we retry and what we shouldn't

Reply:
Since we only run this check for HTTP services, I think we're okay retrying on any GetAsync exception, particularly since we'll eventually timeout and show the endpoint.